torchbenchmark/models/maskrcnn/__init__.py

```diff
@@ -0,0 +1,109 @@
+}
+
+def _collate_fn(batch):
+    return tuple(zip(*batch))
+
+def _prefetch(loader, device):
+    items = []
+    for images, targets in loader:
+        images = list(image.to(device) for image in images)
+        targets = [{k: v.to(device) for k, v in t.items()} for t in targets]
+        items.append((images, targets))
+    return items
+
+class Model(BenchmarkModel):
+    task = COMPUTER_VISION.DETECTION
+
+    def __init__(self, device=None, jit=False, train_bs=4, eval_bs=4):
+        self.device = device
+        self.jit = jit
+        self.model = torchvision.models.detection.maskrcnn_resnet50_fpn(pretrained=True).to(self.device)
+
+        # setup optimizer
+        lr = 0.02
+        momentum = 0.9
+        weight_decay = 1e-4
+        params = [p for p in self.model.parameters() if p.requires_grad]
+        self.optimizer = torch.optim.SGD(params, lr=lr, momentum=momentum, weight_decay=weight_decay)
+
+        self.train_bs = train_bs
+        self.eval_bs = eval_bs
+        t = [ConvertCocoPolysToMask()]
+        transforms = Compose(t)
+        dataset = CocoDetection(root=os.path.join(DATA_DIR, COCO_DATA[COCO_DATA_KEY][0]),
+                                annFile=os.path.join(DATA_DIR, COCO_DATA[COCO_DATA_KEY][1]),
+                                transforms=transforms)
+        sampler = torch.utils.data.SequentialSampler(dataset)
+
+        self.eval_data_loader = _prefetch(torch.utils.data.DataLoader(dataset, batch_size=eval_bs,
+                                                                      sampler=sampler,
+                                                                      collate_fn=_collate_fn), self.device)
+        self.train_data_loader = _prefetch(torch.utils.data.DataLoader(dataset, batch_size=train_bs,
+                                                                       sampler=sampler,
+                                                                       collate_fn=_collate_fn), self.device)
+
+
+    def get_module(self):
+        for (example_inputs, example_targets) in self.train_loader:
+            return self.model, (example_inputs, example_targets, )
+
+    def train(self, niter=1):
+        if self.jit:
+            return NotImplementedError("JIT is not supported by this model")
+        if not self.device == "cuda":
+            return NotImplementedError("CPU is not supported by this model")
+        self.model.train()
+        for _, (images, targets) in zip(range(niter), self.train_data_loader):
+            # images = list(image.to(self.device) for image in images)
+            # targets = [{k: v.to(self.device) for k, v in t.items()} for t in targets]
+            loss_dict = self.model(images, targets)
+            losses = sum(loss for loss in loss_dict.values())
+            self.optimizer.zero_grad()
+            losses.backward()
+            self.optimizer.step()
+
+    def eval(self, niter=1):
```

Comment:
@xuzhao9 I'm a bit concerned we seem to be getting different batches out of `eval_data_loader` every time we run `eval`? Since maskrcnn seems to be data-dependent and bath sizes are so small, we might get running times for eval that vary greatly from one run to another (i.e. noise).
Is this how we usually use data loaders?

Reply:
@Krovatkin Could you please explain why we are getting different batches out of `eval_data_loader` every time? I am thinking we are always getting the first batch because `self.eval_data_loader` is actually a python list that always start from index `0`. Even if it is a data loader object, I think in a for loop it also always start from the first batch?